coderd/notifications/metrics_test.go

```diff
@@ -0,0 +1,342 @@
+
+			if debug {
+				t.Logf("coderd_notifications_dispatched_count == 1: %v", metric.Counter.GetValue())
+			}
+
+			// Only 1 message will be dispatched successfully
+			return metric.Counter.GetValue() == 1
+		},
+		"coderd_notifications_temporary_failures_count": func(series string, metric *dto.Metric) bool {
+			assert.Truef(t, hasMatchingFingerprint(metric, methodTemplateFP), "found unexpected series %q", series)
+
+			if debug {
+				t.Logf("coderd_notifications_temporary_failures_count == %v: %v", maxAttempts-1, metric.Counter.GetValue())
+			}
+
+			// 2 temp failures, on the 3rd it'll be marked permanent failure
+			return metric.Counter.GetValue() == maxAttempts-1
+		},
+		"coderd_notifications_permanent_failures_count": func(series string, metric *dto.Metric) bool {
+			assert.Truef(t, hasMatchingFingerprint(metric, methodTemplateFP), "found unexpected series %q", series)
+
+			if debug {
+				t.Logf("coderd_notifications_permanent_failures_count == 1: %v", metric.Counter.GetValue())
+			}
+
+			// 1 permanent failure after retries exhausted
+			return metric.Counter.GetValue() == 1
+		},
+		"coderd_notifications_retry_count": func(series string, metric *dto.Metric) bool {
+			assert.Truef(t, hasMatchingFingerprint(metric, methodTemplateFP), "found unexpected series %q", series)
+
+			if debug {
+				t.Logf("coderd_notifications_retry_count == %v: %v", maxAttempts-1, metric.Counter.GetValue())
+			}
+
+			// 1 original attempts + 2 retries = maxAttempts
+			return metric.Counter.GetValue() == maxAttempts-1
+		},
+		"coderd_notifications_queued_seconds": func(series string, metric *dto.Metric) bool {
+			assert.Truef(t, hasMatchingFingerprint(metric, methodFP), "found unexpected series %q", series)
+
+			if debug {
+				t.Logf("coderd_notifications_queued_seconds > 0: %v", metric.Histogram.GetSampleSum())
+			}
+
+			// Notifications will queue for a non-zero amount of time.
+			return metric.Histogram.GetSampleSum() > 0
+		},
+		"coderd_notifications_dispatcher_send_seconds": func(series string, metric *dto.Metric) bool {
+			assert.Truef(t, hasMatchingFingerprint(metric, methodFP), "found unexpected series %q", series)
+
+			if debug {
+				t.Logf("coderd_notifications_dispatcher_send_seconds > 0: %v", metric.Histogram.GetSampleSum())
+			}
+
+			// Dispatches should take a non-zero amount of time.
+			return metric.Histogram.GetSampleSum() > 0
+		},
+		"coderd_notifications_pending_updates": func(series string, metric *dto.Metric) bool {
+			// This is a gauge - so we just have to prove it was _once_ set.
+			// See TestPendingUpdatesMetrics for a more precise test.
+			if !seenPendingUpdates {
+				seenPendingUpdates = metric.Gauge.GetValue() > 0
+			}
+
+			if debug {
+				t.Logf("coderd_notifications_pending_updates: %v", seenPendingUpdates)
+			}
+
+			return seenPendingUpdates
+		},
+	}
+
+	// then
+	require.Eventually(t, func() bool {
```

Comment:
Note: Polling designs are often flakey, it would be nice if we could refactor it. Something to keep in mind for the future.

Reply:
Indeed, I have https://github.com/coder/team-coconut/issues/6 scheduled.